### PR TITLE
feat(one-signal): add override to use include_external_user_ids options

### LINKED ISSUE #3479: 🚀 Feature: OneSignal provider overrides to send push notification with external_user_id
### 🔖 Feature description

Hi, I have seen the onesignal feature provided in latest `0.15.0` release but it seems for now novu is only using `player_id` to target the push notification. I think the `external_user_id` could be easier to use in some scenario (OneSignal handle the device and user relationship instead of Novu). Would it be better if there's some override setting specific for OneSignal?

### 🎤 Why is this feature needed ?

In my use-case there's still some limitation on how to manage `deviceTokens` with Novu `setCredentials` method so it would be easier if I can use overrides value to target the push notification with `external_user_id` instead of `player_id`

### ✌️ How do you aim to achieve this?

A new overrides key specific to OneSignal provider should be used and if that key is available, do not use `include_player_ids` args in OneSignal [createNotification](https://documentation.onesignal.com/reference/create-notification) method and use `include_external_user_ids` and `channel_for_external_user_ids='push'` instead

### 🔄️ Additional Information

_No response_

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Contributing Guidelines](https://github.com/novuhq/novu/blob/main/CONTRIBUTING.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by im-perativa**
### What change does this PR introduce?

Add new override key specific to OneSignal provider to use `include_external_user_ids` instead of `include_player_ids` in OneSignal [createNotification](https://documentation.onesignal.com/reference/create-notification) method 

### Why was this change needed?

Closes #3479

### CODE REVIEW

File: packages/stateless/src/lib/provider/provider.interface.ts [hunk 1]
```diff
@@ -64,6 +64,9 @@ export interface IPushOptions {
       };
     };
     fcmOptions?: { analyticsLabel?: string };
+    'one-signal'?: {
```
Comment (p-fernandez) [inline packages/stateless/src/lib/provider/provider.interface.ts:67]: Either this or the `ITriggerOverrides` in `events.interface` is wrong.
We should add tests to the provider to double check the right implementation.

File: packages/stateless/src/lib/provider/provider.interface.ts [hunk 2]
```diff
@@ -64,6 +64,9 @@ export interface IPushOptions {
...
+      externalIds?: string[];
```
Comment (p-fernandez) [inline packages/stateless/src/lib/provider/provider.interface.ts:68]:
```suggestion
      externalUserIds?: string[];
```

**Review by p-fernandez — CHANGES_REQUESTED** [review]
Thanks for your contribution!

File: packages/stateless/src/lib/provider/provider.interface.ts (hunk 1 above)
Comment (im-perativa) [reply packages/stateless/src/lib/provider/provider.interface.ts:67]: Thanks for the feedback. Forgot about that one. Will try to fix and include the docs tomorrow

File: apps/worker/src/app/workflow/usecases/send-message/send-message-push.usecase.ts [hunk 3]
```diff
@@ -155,7 +155,7 @@ export class SendMessagePush extends SendMessageBase {
         continue;
       }
 
-      const overrides = command.overrides[integration.providerId] || {};
+      const overrides = Object.assign(command.overrides, command.overrides[integration.providerId]);
```
Comment (p-fernandez) [inline apps/worker/src/app/workflow/usecases/send-message/send-message-push.usecase.ts:158]: What's the reason behind this change? I am afraid that it can led to buggy behaviour.

File: providers/one-signal/src/lib/one-signal.provider.ts [hunk 4]
```diff
@@ -29,7 +29,12 @@ export class OneSignalPushProvider implements IPushProvider {
     const { sound, badge, ...overrides } = options.overrides ?? {};
 
     const res = await this.oneSignal.createNotification({
-      include_player_ids: options.target,
+      ...(overrides['one-signal']?.externalUserIds !== undefined
+        ? {
+            include_external_user_ids: overrides['one-signal'].externalUserIds,
+            channel_for_external_user_ids: 'push',
+          }
+        : { include_player_ids: options.target }),
```
Comment (p-fernandez) [inline providers/one-signal/src/lib/one-signal.provider.ts:37]:
```suggestion
   const { externalUserIds } =  overrides['one-signal'] || {};
   const res = await this.oneSignal.createNotification({ 
      ...(externalUserIds !== undefined
        ? {
            include_external_user_ids: externalUserIds,
            channel_for_external_user_ids: 'push',
          }
        : { include_player_ids: options.target }),
```
We can reuse the extraction of the `externalUserIds`.
Any way of doing it should be enough:
```
   const externalUserIds =  overrides['one-signal']?.externalUserIds;
```

**Comment by im-perativa** [comment]
> @im-perativa I saw you requested a re-review but there is an open pending question to address. 🙏🏻

Sorry I didn't know my previous comment on your question was still on pending state 😅

File: docs/docs/channels/push/one-signal.md [hunk 5]
```diff
@@ -30,6 +30,8 @@ await novu.subscribers.setCredentials('subscriberId', PushProviderIdEnum.OneSign
 });
 ```
 
+If your user have multiple devices and you want to leave the device-to-user relationship to OneSignal, you can use `externalUserIds` override to send push notifications to target [external_user_ids](https://documentation.onesignal.com/docs/external-user-ids) and skip the registration of [player_id](https://documentation.onesignal.com/docs/users#player-id) as `deviceToken` in Novu. You must set `external_user_ids` with OneSignal [setExternalUserId](https://documentation.onesignal.com/docs/external-user-ids#setexternaluserid-method) method to associate multiple `player_id` with `external_user_id` in your client first.
+
```
[the remark this answers is not in the document]
Comment (im-perativa) [reply docs/docs/channels/push/one-signal.md:34]: @L-U-C-K-Y sure, I haven't got the time to read the full deprecation notice thoroughly but I think we should update this MR first

[the remark this answers is not in the document]
Comment (L-U-C-K-Y) [reply docs/docs/channels/push/one-signal.md:34]: @im-perativa I agree, let's upgrade to latest OneSignal SDK and rename to external_id, I think this PR in general would be very useful. Not sure if I can work on it this week, but for sure I can review.

**Comment by scopsy** [comment]
@im-perativa we have recently upgraded to the latest one signal client. I'm closing this PR, as I think it will be easier to just recreate the change instead of merging conflicts. My apologies for this PR has gotten stale from our side 🙏
